Issue #51: Implement Checkstyle report generation script in Groovy

MEZk opened the pull request:
Implemented Checkstyle report generation script in Groovy.

Report folder diff (shell vs Groovy implementation):

![cmp](https://cloud.githubusercontent.com/assets/7242568/20499775/020ce516-b044-11e6-9d5e-99d81bd2e76b.png)

script cli args example:  ```groovy launch -c my_check.xml -p projects-to-test-on.properties```

Should I update README to explain how to use the new script?

rnveach commented:
@MEZk I assume this is for issue 51, not 52?
I think we should have a readme for it. I'm not sure if it should be in the main readme or its own, as we will want to directly link to it for newcomers.
Also remember some people don't know groovy, like me, and they may be more familar with Java since that is what the main project is written in.

MEZk commented:
@rnveach 
Yes, it is for #51 ) Fixed.

>I think we should have a readme for it. I'm not sure if it should be in the main readme or its own, as we will want to directly link to it for newcomers.

@romani 
What is your opinion?

>Also remember some people don't know groovy, like me, and they may be more familar with Java since that is what the main project is written in.

Groovy is similar to java and run in JVM. It is more readable and easy to support for newbies. I also do not know Groovy at all, but my experience in Java made it easy for me to implement the script. Pure Java is to heavy for scripting, and even though Groovy is running in JVM, it has lightweight scripting support.

rnveach commented:
@MEZk 
I'm confused. Isn't this issue suppose to be a full automation script of regression like my script, `launch_diff.sh`?
This just looks more like `launch.sh`. Are we doing both in groovy?

romani commented:
I presume that this is a just a first milestone on #51 - change launch_sh to groovy. 
Next milestone is finally resolve #51 and have groovy script to generate diff report too.
@MEZk , please confirm your intentions.

MEZk commented:
@rnveach @romani 
This is just the first step. Next step will be the Groovy script which uses launch.groovy to generate reports and get diff based on them. After the merge of this PR I am going to change CIs yml matrix to run the new script in Groovy.

rnveach commented:
> Next step will be the Groovy script which uses launch.groovy to generate reports and get diff based on them.

@romani @MEZk 
Just to be clear, `launch_diff` doesn't call `launch`.

Each project gets its own separate maven run, which is not what happens in `launch`. In `launch` everything is packed into one directory and JXR is run in one pass.
There are very good reasons for why my script didn't follow suite.

Putting too many projects in one run, requires too much memory. We also need more memory for violations and some checks like indentation put out alot of violations. I was always required to give indentation 4 gigs of memory to complete all projects, even though each project was run separate.
Also if you pack all the projects together, you won't get 1 HTML report per project unless you modify the resulting `checkstyle-result.xml` that the diff report needs. All the projects will be combined into one HTML, and could be huge and slow to load in the browser.

Another thing my script did, was verify that the checkstyle version installed matches the version specified in the [POM](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/pom.xml#L15).
If they aren't in synch, the checkstyle version run is usually an old version and there is no clear way to notice this unless you are expecting differences. This caused almost quite a few invalid regression reports for me until I found and added the check.
You can't really do this unless it is one full suite or passing data back and forth.

I am not saying we can't convert `launch` and use it, but I don't think it can be used for the full regression suite.

MEZk commented:
@rnveach @romani 

>Each project gets its own separate maven run, which is not what happens in launch

We can add an option which specifies the project to generate report for. It will have higher priority than "-p"  and run the script only for one project.

>Putting too many projects in one run, requires too much memory. 

Yes. CIs will not allow us to use so much memory. And it is a problem, but if we split report generation into steps (each for one project) it will help.

romani commented:
This is first step to reusable code. Eventually groovy script will do all that possible by both of shell scripts. And we will have one script that we need to support.

@rnveach , please join me in review of this PR.

romani commented on checkstyle-tester/launch.groovy:
```diff
@@ -0,0 +1,191 @@
+                cloneRepository(repoName, repoType, repoUrl, commitId, srcDir)
+            }
+    }
+    runMavenExecution(srcDir, excludes, checkstyleConfig)
```
please be consistent in sequence of arguments. I like this sequence , so please fix "+def generateCheckstyleReport(checkstyleConfig, srcDir, projectsToTestOn) {"
to be 
"def generateCheckstyleReport(srcDir, projectsToTestOn, checkstyleConfig) {"

romani commented on checkstyle-tester/launch.groovy:
```diff
@@ -0,0 +1,191 @@
+def postProcessCheckstyleReport() {
+    println 'linking report to index.html'
+    new File('target/site/index.html').renameTo 'target/site/_index.html'
+    Files.createLink(Paths.get("target/site/index.html"), Paths.get("target/site/checkstyle.html"))
```
I am not sure this will work for Windows, @rnveach please confirm.
we can use "mv"/"rename".

rnveach replied:
It should work. http://stackoverflow.com/questions/12509132/how-to-force-java-7-to-create-directory-symlinks-on-windows-instead-of-file-s?rq=1
I've played with another utility that creates symlinks on Windows 7, so it is supported, just not easily accessible.
http://www.howtogeek.com/howto/16226/complete-guide-to-symbolic-links-symlinks-on-windows-or-linux/

rename in windows is: `ren`. mv in windows is: `move`.

romani commented on checkstyle-tester/launch.groovy:
```diff
@@ -0,0 +1,191 @@
+def removeNonReferencedXrefFiles() {
+    println 'Removing non refernced xref files in report ...'
+
+    def linesFromIndexHtml = Files.readAllLines(Paths.get('target/site/index.html'))
```
make a function from this block.

romani commented on checkstyle-tester/launch.groovy:
```diff
@@ -0,0 +1,191 @@
+def removeEmptyDirectories() {
+    // we use unix commands to avoid problems with concurrency modification of collection
+    // (removing and iteration over collection) in Groovy
+    executeCmd('find . -type d -empty -delete', new File('./target/site/xref'))
```
non Windows friendly. But might be ok for first iteration.

romani commented on checkstyle-tester/launch.groovy:
```diff
@@ -0,0 +1,191 @@
+    proc.waitFor()
+    if (proc.exitValue() != 0) {
+        println "Error: ${proc.err.text}!"
+        System.exit(-1)
```
please throw exception there.

rnveach commented:
Github is weird. I finished my review now, but because some were made long ago, it put the 2 review items and comments up near the top of this thread.

MEZk commented:
@romani

> I think it make sense to use "public static void main(args) {"

Done.

> is it really required ?

No. Removed.

> please be consistent in sequence of arguments. I like this sequence , so please fix "+def generateCheckstyleReport(checkstyleConfig, srcDir, projectsToTestOn) {"
to be
"def generateCheckstyleReport(srcDir, projectsToTestOn, checkstyleConfig) {"

Put srcDir and other paths which are used by generateCheckstyleReport inside generateCheckstyleReport

> make a function from this block.

Done.

> non Windows friendly.

Done. Implemented recursive method to delete empty directories.

MEZk commented:
@rnveach

> Since we are downloading the repos and not cleaning them out, for long term users who will use this for years to come, you should put in some git fetch and reset commands like launch_update. Just so we are always running on the project's latest code, unless a specific commit is set.

Contributor can use 'master'/'HEAD' for git or 'default' for mercurial tags in projects-to-test-on.properties to force the script to download the lates sources. Maybe it will be better to add description in README.

> I recommend putting all downloaded repos in a separate directory not in src/main/java and then copying them over for the run.
Any new directories added will have to be added to gitignore.

Done.

> Will I have to specify the one project to run? I would dislike having to redo the loading/parsing of the projects to test on outside of this script. Can't we just make this script run all the projects in a single/separate run?

Yes. Done.

> do we need to sanitize directory name $repoName for invalid characters?
Windows users have some characters that can't be in directory names.
Can wait when we make this more windows friendly.

Paths.get() will throw IllegalPathException in case the path is not correct. 

> Maybe verify if current commit id is equal to commitId? Then we can avoid doing a useless reset.

Done.

MEZk commented:
@romani @rnveach 
Now the script copies the full content of 'target' directory ('site' folder, checkstyle-result.xml, checkstyle-checker.xml) into reports directory. However, diff tool requires only checkstyle-result.xml. Nevertheless, I decided to leave the whole content as contributer may need to analyze report just 'after' fix without diff. But the size of such reports may be too big. What is your opinion? Should we have an option to generate just checkstyle-result.xml?

romani commented:
Diff report and simple html report are completely different modes. Both are required for now, in future smth could change as checkstyle could generate html report itself.

romani commented:
It is fine for first step , very good step forward.

MEZk commented:
@romani 
Did you merge changes? I do not see the script in master branch.

romani commented:
Merged.
Sorry.
